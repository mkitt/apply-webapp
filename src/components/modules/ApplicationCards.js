// @flow
import React, { type Node } from 'react'
import type { Location } from 'react-router-dom'
import glamorous from 'glamorous'
import { map } from 'ramda'
import ApplicationCard from './ApplicationCard'
import Loader from '../blocks/Loader'
import AppContainment from '../elements/AppContainment'
import type { Application } from '../../types/Application'

type Props = {
  children: Node,
  allApplications: Array<Application>,
  refetch: () => {},
  location: Location,
}

export default class extends React.PureComponent<Props> {
  componentWillReceiveProps(nextProps: Props) {
    if (this.props.location.key !== nextProps.location.key) {
      this.props.refetch()
    }
  }

  render() {
    const { allApplications, children } = this.props
    return (
      <AppContainment>
        <Cards>
          {map(application => (
            <ApplicationCard
              applicant={application.applicant}
              id={application.id}
              key={application.id}
            />
          ), allApplications)}
        </Cards>
        {children}
      </AppContainment>
    )
  }
}

export const ComponentLoader = () => (
  <Loader>Loading Applications...</Loader>
)

// -------------------------------------

const Cards = glamorous.section({
  display: 'grid',
  gridTemplateColumns: '1fr',
  gridGap: 16,
  '> *': {
    minWidth: '100%',
  },
  '@media (min-width: 41em)': {
    gridTemplateColumns: '1fr 1fr',
    '> *': {
      minWidth: 320,
    },
  },
  '@media (min-width: 62em)': {
    gridTemplateColumns: '1fr 1fr 1fr',
  },
  '@media (min-width: 83em)': {
    gridTemplateColumns: '1fr 1fr 1fr 1fr',
  },
})
