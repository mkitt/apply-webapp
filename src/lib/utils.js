// @flow
import { __, addIndex, always, and, complement, either, gt, isEmpty, isNil, map } from 'ramda'
import { withProps } from 'recompose'

// -------------------------------------
// Environments
export const isDev = always(process.env.NODE_ENV === 'development')

// -------------------------------------
// Types
export const isNotNil = complement(isNil)
export const isNotEmpty = complement(isEmpty)
export const isEmptyOrNil = either(isNil, isEmpty)
export const isNotEmptyAndNotNil = and(isNotNil, isNotEmpty)

// -------------------------------------
// Relation
export const gt0 = gt(__, 0)

// -------------------------------------
// Lists
export const nool = always([])()
export const mapIndexed = addIndex(map)

// -------------------------------------
// Objects
export const noob = always({})()

// -------------------------------------
// Functions
export const noop = () => {}

export const log = withProps(props => (
  // eslint-disable-next-line no-console
  console.log(props.displayName || 'üëç', props) && props
))

export const warn = withProps(props => (
  // eslint-disable-next-line no-console
  console.warn(props.displayName || '‚ö†Ô∏è', props) && props
))
