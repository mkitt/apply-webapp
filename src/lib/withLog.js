import { withProps } from 'recompose'

export default withProps(props => (
  // eslint-disable-next-line
  console.log(props.logName || props.displayName || '👍', props) && props
))
