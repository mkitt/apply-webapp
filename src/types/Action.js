// @flow
import type { Applicant } from './Applicant'

export type Action = {
  applicants: Array<Applicant>,
  id: ID,
  label: string,
  message: string,
  name: string,
}
