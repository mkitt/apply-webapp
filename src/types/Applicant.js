// @flow
import { camelize } from 'humps'
import { compose, prop, split, head, last } from 'ramda'
import type { Action } from './Action'
import type { Application } from './Application'

export default undefined

export type Applicant = {
  action: ?Action,
  applications: Array<Application>,
  birthDate: ?string,
  createdAt: DateTime,
  email: string,
  github: string,
  id: ID,
  name: string,
  referredBy: ?string,
  resume: ?string,
  startDate: ?DateTime,
  status: ?string,
  updatedAt: DateTime,
  // derived data...
  firstName: string,
  lastName: string,
}

// -------------------------------------
// Derived helpers

export const deriveFirstName = compose(
  head,
  split(' '),
  prop('name'),
)

export const deriveLastName = compose(
  last,
  split(' '),
  prop('name'),
)

// -------------------------------------
// For testing content
export const stub = (props: Applicant) => {
  const firstName = (props != null && props.firstName) || 'Kenny'
  const lastName = (props != null && props.lastName) || 'Bania'
  const camel = camelize(`${firstName} ${lastName}`)
  const applicant = {
    id: camel,
    action: null,
    applications: [],
    birthDate: '06/06/2006',
    createdAt: '2010-10-31T12:12:12.000Z',
    email: 'kenny.bania@example.com',
    firstName: 'Kenny',
    github: camel,
    lastName,
    referredBy: 'Simon Taranto',
    resume: 'https://www.resume.com',
    startDate: null,
    status: 'pending',
    updatedAt: '2010-12-24T24:24:24.000Z',
  }
  return {
    ...applicant,
    ...props,
  }
}
