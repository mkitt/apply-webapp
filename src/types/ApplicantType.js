// @flow
export type ApplicantType = {
  applyAction: string,
  applyStatus: string,
  birthdate: string,
  createdAt: string,
  email: string,
  firstName: string,
  github: string,
  id: string,
  lastName: string,
  referredBy: string,
  resume: string,
  scoreLogicEvaluation: number,
  scoreOnlineLogicTest: number,
  scoreValuesEvaluation: number,
  startDate: string,
  updatedAt: string,
}
