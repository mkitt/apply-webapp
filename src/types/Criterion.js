// @flow
import { camelize } from 'humps'
import { compose, prop } from 'ramda'
import type { Evaluation } from './Evaluation'

export type Criterion = {
  id: ID,
  createdAt: DateTime,
  evaluation: ?Evaluation,
  title: string,
  options: Array<string>,
  questions: Array<string>,
  updatedAt: DateTime,
  // derived...
  name: string,
}

// -------------------------------------
// Derived helpers

export const deriveName = compose(
  camelize,
  prop('title'),
)

// -------------------------------------
// For testing content
export const stub = (props: Criterion) => {
  const title = (props != null && props.title) || 'Problem Solving'
  const camel = camelize(title)
  return {
    id: camel,
    title,
    name: camel,
    options: [
      `${title} Option 3`,
      `${title} Option 2`,
      `${title} Option 1`,
    ],
    questions: [
      `${title} Question 1`,
      `${title} Question 2`,
      `${title} Question 3`,
    ],
    ...props,
  }
}

export default undefined
