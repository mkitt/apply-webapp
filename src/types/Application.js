// @flow
import { both, compose, either, prop, propSatisfies, replace, test } from 'ramda'
import { capitalize, gt0 } from '../lib/utils'
import type { Applicant } from './Applicant'
import type { Evaluator } from './Evaluator'
import type { Evaluation } from './Evaluation'

export type Application = {
  id: ID,
  applicant: Applicant,
  complete: boolean,
  evaluations: Array<Evaluation>,
  evaluators: ?Array<Evaluator>,
  resume: ?string,
  scoreLogicEvaluation: ?Int,
  scoreOnlineLogicTest: ?Int,
  scoreValuesEvaluation: ?Int,
  status: string,
  // derived...
  isStatusMutatable: boolean,
  statusLabel: string,
}

// -------------------------------------
// Derived helpers

/*
export type StatusType = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14
export const STATUS = Object.freeze({
  pending: 0,
  needs_initial_evaluation_scores: 1,
  needs_rejected_at_initial_evaluation_notification: 2,
  rejected_at_initial_evaluation: 3,
  needs_to_schedule_interview: 4,
  needs_interview_scores: 5,
  needs_rejected_at_interview_notification: 6,
  rejected_at_interview: 7,
  needs_logic_evaluation_scores: 8,
  needs_rejected_at_logic_evaluation_notification: 9,
  rejected_at_logic_evaluation: 10,
  needs_invitation: 11,
  needs_invitation_response: 12,
  declined_invitation: 13,
  accepted_invitation: 14,
})
*/

export const needsSendStatus = compose(
  test(/^needs_rej|needs_invitation$/),
  prop('status'),
)

export const needsSubmitEvaluation = compose(
  test(/_evaluation_scores$|^needs_interview/),
  prop('status'),
)

export const deriveIsStatusMutatable = either(
  propSatisfies(test(/^needs_rej|needs_invitation$|_evaluation_scores$/), 'status'),
  both(
    propSatisfies(test(/^needs_interview/), 'status'),
    propSatisfies(gt0, 'scoreOnlineLogicTest'),
  ),
)

export const deriveStatusLabel = compose(
  capitalize,
  replace(/_/gi, ' '),
  prop('status'),
)

// -------------------------------------
// For testing content
export const stub = (props: Application) => {
  const application = {
    id: '1',
    complete: false,
    evaluations: [],
    evaluators: null,
    isStatusMutatable: false,
    resume: 'uploads/asset/application-resume.pdf',
    scoreLogicEvaluation: 20,
    scoreOnlineLogicTest: 40,
    scoreValuesEvaluation: 20,
    showScores: true,
    status: 'pending',
    statusLabel: 'Pending',
  }
  return {
    ...application,
    ...props,
  }
}

export default undefined
