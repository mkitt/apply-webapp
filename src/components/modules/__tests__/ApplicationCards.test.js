import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import ApplicationCards from '../ApplicationCards'
import { stub } from '../../../types/Application'
import { stub as stubApplicant } from '../../../types/Applicant'
import { noop } from '../../../lib/utils'

const allApplications = [
  stub({ applicant: stubApplicant() }),
  stub({ id: '22', applicant: stubApplicant({ firstName: 'Robert', lastName: 'Sacamano', id: '2' }) }),
]

describe('<ApplicantCard />', () => {
  it('shows the correct snapshot tree for the <ApplicationCards> module', () => {
    const component = (
      <MemoryRouter>
        <ApplicationCards
          allApplications={allApplications}
          location={{ key: '1' }}
          refetch={{ noop }}
        />
      </MemoryRouter>
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
