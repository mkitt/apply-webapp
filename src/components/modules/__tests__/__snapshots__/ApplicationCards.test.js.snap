// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApplicantCard /> shows the correct snapshot tree for the <ApplicationCards> module 1`] = `
.css-0,
[data-css-0] {
  position: relative;
  max-width: 1392px;
  margin-right: auto;
  margin-left: auto;
  padding-right: 8px;
  padding-bottom: 16px;
  padding-left: 8px;
}

.css-1,
[data-css-1] {
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 16px;
  -ms-grid-template-columns: 1fr;
  -ms-grid-gap: 16px;
}

.css-1> *,
[data-css-1]> * {
  min-width: 100%;
}

.css-2,
[data-css-2] {
  position: relative;
  min-width: 320px;
  height: 408px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.2);
}

.css-3,
[data-css-3] {
  height: 4px;
  background-color: #4d4d4d;
}

.css-4,
[data-css-4] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  height: 160px;
  padding-right: 24px;
  padding-bottom: 16px;
  padding-left: 24px;
  background-color: #e5e5e5;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-5,
[data-css-5] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  font-size: 34px;
  font-weight: 200;
  color: #808080;
  margin-top: auto;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-6,
[data-css-6] {
  margin-top: 24px;
  padding-right: 24px;
  padding-left: 24px;
  color: #4d4d4d;
}

.css-7,
[data-css-7] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  margin-top: 8px;
  margin-bottom: 8px;
  color: currentColor;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-8,
[data-css-8] {
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  color: currentColor;
}

.css-9,
[data-css-9] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  margin-left: 0;
  font-size: 16px;
  font-weight: 200;
  color: currentColor;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-10,
[data-css-10] {
  position: relative;
  z-index: 2;
  color: #35c5fc;
  text-decoration: none;
  transition: color 200ms;
  -webkit-transition: color 200ms;
  -moz-transition: color 200ms;
}

.css-11,
[data-css-11] {
  position: relative;
  z-index: 2;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  margin-top: 24px;
  margin-bottom: 8px;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-12,
[data-css-12] {
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  color: #999;
}

.css-13,
[data-css-13] {
  margin-top: 8px;
  margin-left: 0;
}

.css-14,
[data-css-14] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  margin-top: 8px;
  font-size: 14px;
  font-weight: 400;
  text-transform: uppercase;
  color: #4d4d4d;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-15,
[data-css-15] {
  margin-right: 8px;
}

.css-16,
[data-css-16] {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  background-color: rgba(0, 0, 0, 0);
}

@media (min-width: 41em) {
  .css-1,
  [data-css-1] {
    grid-template-columns: 1fr 1fr;
    -ms-grid-template-columns: 1fr 1fr;
  }

  .css-1> *,
  [data-css-1]> * {
    min-width: 320px;
  }
}

@media (min-width: 62em) {
  .css-1,
  [data-css-1] {
    grid-template-columns: 1fr 1fr 1fr;
    -ms-grid-template-columns: 1fr 1fr 1fr;
  }
}

@media (min-width: 83em) {
  .css-1,
  [data-css-1] {
    grid-template-columns: 1fr 1fr 1fr 1fr;
    -ms-grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}

<div
  className="css-0"
>
  <section
    className="css-1"
  >
    <div
      className="css-2"
    >
      <div
        className="css-3"
      />
      <header
        className="css-4"
      >
        <h2
          className="css-5"
        >
          <span>
            Kenny
             
          </span>
          <span>
            Bania
          </span>
        </h2>
      </header>
      <div
        className="css-6"
      >
        <dl
          className="css-7"
        >
          <dt
            className="css-8"
          >
            Email Address
          </dt>
          <dd
            className="css-9"
          >
            <a
              className="css-10"
              href="mailto:kenny.bania@example.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              kenny.bania@example.com
            </a>
          </dd>
        </dl>
        <dl
          className="css-7"
        >
          <dt
            className="css-8"
          >
            Login
          </dt>
          <dd
            className="css-9"
          >
            <a
              className="css-10"
              href="https://github.com/kennyBania"
              rel="noopener noreferrer"
              target="_blank"
            >
              kennyBania
            </a>
          </dd>
        </dl>
        <dl
          className="css-11"
        >
          <dt
            className="css-12"
          >
            Application Status
          </dt>
          <dd
            className="css-13"
          >
            <span
              className="css-14"
            >
              <svg
                className="css-15"
                height={32}
                viewBox="0 0 24 24"
                width={32}
              >
                <g
                  fill="currentColor"
                >
                  <path
                    d="M24,0 C24,0 23.4836048,9.80252996 16.4820958,15.2719707 L16.4128111,19.0346251 L11.4820194,24 L11.4746834,20.2882431 L9.63105705,18.6292449 C8.29671435,22.7691304 1.26439406,22.7691304 1.26439406,22.7691304 C1.26439406,22.7691304 1.26439406,16.3396736 5.58319762,14.4450082 L3.94220876,12.6079505 L-2.84217094e-14,12.6079505 L5.07440302,7.66843342 L8.7285621,7.58627726 C14.4472436,0.671411685 24,0 24,0 Z M16.425,10.1644246 C17.8333261,10.1644246 18.975,9.01830529 18.975,7.60449544 C18.975,6.19068562 17.8333261,5.0445663 16.425,5.0445663 C15.0166739,5.0445663 13.875,6.19068562 13.875,7.60449544 C13.875,9.01830529 15.0166739,10.1644246 16.425,10.1644246 Z"
                  />
                </g>
              </svg>
              Pending
            </span>
          </dd>
        </dl>
      </div>
      <a
        className="css-16"
        href="/application/1"
        onClick={[Function]}
      />
    </div>
    <div
      className="css-2"
    >
      <div
        className="css-3"
      />
      <header
        className="css-4"
      >
        <h2
          className="css-5"
        >
          <span>
            Robert
             
          </span>
          <span>
            Sacamano
          </span>
        </h2>
      </header>
      <div
        className="css-6"
      >
        <dl
          className="css-7"
        >
          <dt
            className="css-8"
          >
            Email Address
          </dt>
          <dd
            className="css-9"
          >
            <a
              className="css-10"
              href="mailto:kenny.bania@example.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              kenny.bania@example.com
            </a>
          </dd>
        </dl>
        <dl
          className="css-7"
        >
          <dt
            className="css-8"
          >
            Login
          </dt>
          <dd
            className="css-9"
          >
            <a
              className="css-10"
              href="https://github.com/robertSacamano"
              rel="noopener noreferrer"
              target="_blank"
            >
              robertSacamano
            </a>
          </dd>
        </dl>
        <dl
          className="css-11"
        >
          <dt
            className="css-12"
          >
            Application Status
          </dt>
          <dd
            className="css-13"
          >
            <span
              className="css-14"
            >
              <svg
                className="css-15"
                height={32}
                viewBox="0 0 24 24"
                width={32}
              >
                <g
                  fill="currentColor"
                >
                  <path
                    d="M24,0 C24,0 23.4836048,9.80252996 16.4820958,15.2719707 L16.4128111,19.0346251 L11.4820194,24 L11.4746834,20.2882431 L9.63105705,18.6292449 C8.29671435,22.7691304 1.26439406,22.7691304 1.26439406,22.7691304 C1.26439406,22.7691304 1.26439406,16.3396736 5.58319762,14.4450082 L3.94220876,12.6079505 L-2.84217094e-14,12.6079505 L5.07440302,7.66843342 L8.7285621,7.58627726 C14.4472436,0.671411685 24,0 24,0 Z M16.425,10.1644246 C17.8333261,10.1644246 18.975,9.01830529 18.975,7.60449544 C18.975,6.19068562 17.8333261,5.0445663 16.425,5.0445663 C15.0166739,5.0445663 13.875,6.19068562 13.875,7.60449544 C13.875,9.01830529 15.0166739,10.1644246 16.425,10.1644246 Z"
                  />
                </g>
              </svg>
              Pending
            </span>
          </dd>
        </dl>
      </div>
      <a
        className="css-16"
        href="/application/22"
        onClick={[Function]}
      />
    </div>
  </section>
</div>
`;
